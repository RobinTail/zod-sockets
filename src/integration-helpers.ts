import { createHash } from "node:crypto";
import { range } from "ramda";
import ts from "typescript";
import { z } from "zod";

export const f = ts.factory;
export const exportModifier = [f.createModifier(ts.SyntaxKind.ExportKeyword)];

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const lcFirst = (subject: string) =>
  subject.charAt(0).toLowerCase() + subject.slice(1);

export const makeCleanId = (...args: string[]) =>
  args
    .flatMap((entry) => entry.split(/[^A-Z0-9]/gi)) // split by non-alphanumeric characters
    .flatMap((entry) =>
      // split by sequences of capitalized letters
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    )
    .map(ucFirst)
    .join("");

export const defaultSerializer = (schema: z.ZodTypeAny): string =>
  createHash("sha1").update(JSON.stringify(schema), "utf8").digest("hex");

export const makeEventFnSchema = (
  base: z.AnyZodTuple,
  ack?: z.AnyZodTuple,
  maxOverloads: number = 3,
) => {
  if (!ack) {
    return z.function(base, z.void());
  }
  const fn = z.function(ack, z.void());
  const rest = base._def.rest;
  if (!rest || maxOverloads <= 0) {
    return z.function(z.tuple([...base.items, fn]), z.void());
  }
  const variants = range(0, maxOverloads + 1).map((count) => {
    const items = [...base.items]
      .concat(
        range(1, count + 1).map((index) =>
          rest.describe(`${rest.description || "rest"}${index}`),
        ),
      )
      .concat(fn);
    return z.function(z.tuple(items as z.ZodTupleItems), z.void());
  });
  return z.union(
    variants as unknown as [z.ZodTypeAny, z.ZodTypeAny, ...z.ZodTypeAny[]],
  );
};
/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const tryToTransform = <T>(
  schema: z.ZodEffects<z.ZodTypeAny, T>,
  sample: T,
) => {
  try {
    return typeof schema.parse(sample);
  } catch (e) {
    return undefined;
  }
};
