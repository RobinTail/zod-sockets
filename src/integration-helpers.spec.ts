import { z } from "zod";
import { makeEventFnSchema } from "./integration-helpers";

describe("Integration helpers", () => {
  describe("makeEventFnSchema()", () => {
    test("should simply use base when no ack", () => {
      const base = z.tuple([z.string()]);
      const result = makeEventFnSchema(base);
      expect(result).toMatchSnapshot();
    });

    test("should add ack when no rest in base", () => {
      const base = z.tuple([z.string()]);
      const ack = z.tuple([z.number()]);
      const result = makeEventFnSchema(base, ack);
      expect(result).toMatchSnapshot();
    });

    test("should use overloads when both rest and ack are present", () => {
      const base = z.tuple([z.string()]).rest(z.unknown());
      const ack = z.tuple([z.number()]);
      const result = makeEventFnSchema(base, ack, 2);
      expect(result).toMatchSnapshot();
    });
  });
});
