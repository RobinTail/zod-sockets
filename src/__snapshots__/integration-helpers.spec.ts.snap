// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration helpers > makeEventFnSchema() > should add ack when no rest in base 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "x-brand": "function",
  "x-input": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "prefixItems": [
      {
        "type": "string",
      },
      {
        "x-brand": "function",
        "x-input": {
          "$schema": "https://json-schema.org/draft/2020-12/schema",
          "prefixItems": [
            {
              "type": "number",
            },
          ],
          "type": "array",
        },
        "x-output": {
          "$schema": "https://json-schema.org/draft/2020-12/schema",
        },
      },
    ],
    "type": "array",
  },
  "x-output": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
  },
}
`;

exports[`Integration helpers > makeEventFnSchema() > should simply use base when no ack 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "x-brand": "function",
  "x-input": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "prefixItems": [
      {
        "type": "string",
      },
    ],
    "type": "array",
  },
  "x-output": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
  },
}
`;

exports[`Integration helpers > makeEventFnSchema() > should use overloads when both rest and ack are present 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "anyOf": [
    {
      "x-brand": "function",
      "x-input": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "prefixItems": [
          {
            "type": "string",
          },
          {
            "x-brand": "function",
            "x-input": {
              "$schema": "https://json-schema.org/draft/2020-12/schema",
              "prefixItems": [
                {
                  "type": "number",
                },
              ],
              "type": "array",
            },
            "x-output": {
              "$schema": "https://json-schema.org/draft/2020-12/schema",
            },
          },
        ],
        "type": "array",
      },
      "x-output": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
      },
    },
    {
      "x-brand": "function",
      "x-input": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "prefixItems": [
          {
            "type": "string",
          },
          {
            "description": "rest1",
          },
          {
            "x-brand": "function",
            "x-input": {
              "$schema": "https://json-schema.org/draft/2020-12/schema",
              "prefixItems": [
                {
                  "type": "number",
                },
              ],
              "type": "array",
            },
            "x-output": {
              "$schema": "https://json-schema.org/draft/2020-12/schema",
            },
          },
        ],
        "type": "array",
      },
      "x-output": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
      },
    },
  ],
}
`;
